tern/utils/general.py

```diff
@@ -116,6 +117,102 @@ def get_git_rev_or_version():
+                # get sha info from ref
+                with open(sha_path, 'r') as sha_file:
+                    sha_info = sha_file.readline().strip('\n')
+            # find project name
+            name_info = ''
+            conf = configparser.ConfigParser()
+            conf.read(dot_git_folder_path + 'config', encoding='utf-8')
+            # use remote url for git project name
+            if conf.has_section('remote "origin"'):
+                # url is the first item
+                project_url = conf.options('remote "origin"')[0]
+                name_info = project_url.split('/')[-1]
+            else:
+                name_info = '(not found)'
+            comment_line = 'git project name: ' + name_info +\
+                           ' sha info: ' + sha_info
+        else:
+            comment_line = 'Not a git repository'
+    return comment_line
+
+
+def check_git_tar(path_to_tar):
+    '''Given the path to the tar file, return the name and sha of repository.
+    This function looks into .git folder in the tar file'''
+    comment_line = 'Unable to open tar file'
+    with tarfile.open(path_to_tar, 'r') as tar:
+        names = tar.getnames()
+        # first name is the git repository folder name
+        folder_name = names[0]
+        if folder_name + '/.git' in names:
+            # find sha information
+            head_file = tar.extractfile(folder_name + '/.git/HEAD')
+            # decode bytes
+            ref_info = head_file.readline().decode()
+            ref_relative_path = ref_info.split(' ', 1)[1].strip('\n')
+            sha_path = folder_name + '/.git/' + ref_relative_path
+            sha_file = tar.extractfile(sha_path)
+            sha_info = sha_file.readline().decode().strip('\n')
+            # find project name
+            config_file = tar.extractfile(folder_name + '/.git/config')
+            found_url = False
+            for line in config_file:
+                line = line.decode()
+                # match first url line
+                if line.startswith('url'):
+                    found_url = True
+                    name_info = line.split('/')[-1]
+                    break
+            if not found_url:
+                name_info = '(not found)'
+            comment_line = 'git project name: ' + name_info +\
+                ' sha info: ' + sha_info
+        else:
+            comment_line = 'Not a git repository'
+    return comment_line
+
+
+def check_git_src(dockerfile_path, src_path):
+    '''Given the src_path and the dockerfile path, return the git
+    repository name and sha information in the format of string.
+    This function check the type of src_path and use check_git_tar and
+    check_git_folder to handle the src_path'''
+    # get the folder path
+    dockerfile_folder_path = os.path.dirname(dockerfile_path)
+    # get the absolute path of the target folder or tar file
+    src_path = os.path.join(dockerfile_folder_path, src_path)
```

Comment:
What if the potential `git` directory, specified by `src_path`, is not in the same directory as the Dockerfile folder path? Or what if  the Dockerfile is in a few subfolders below? @nishakm not sure how hard we want to look for the source that is being copied in. It's also common to see the source being copied in as just `.` which could be multiple tars or directories. We should probably account for this in some way even if we don't filter through all the directories for potential git information.

Reply:
In general the src file path is the relative path to the Dockerfile path. I tested that it works with the git repo in the subfolders. If the dockerfile is in a few subfolders below, the given dockerfile_path is changed. What we do here is looking around the folder containing dockerfile and its subfolders. So we need to locate the dockerfile in this situation. I am wondering if I have missed some situations.

Reply:
>> In general the src file path is the relative path to the Dockerfile path

This is true in most cases, but we need to account for when it is not true. People do weird things with Dockerfiles :) 

>> If the dockerfile is in a few subfolders below, the given dockerfile_path is changed.

If the Dockerfile is a few subfolders below the git folder then wouldn't `src_path = os.path.join(dockerfile_folder_path, src_path)` give you `dir/dockerfile_folder_below/gitdirectory` even though the git directory could be a directory above in `dir`?